docs/source/daml/intro/4_Transformations.rst

```diff
@@ -76,13 +76,14 @@ Choices in the Ledger Model
 
 In :doc:`1_Token` you learned about the high-level structure of a DAML ledger. With choices and the `exercise` function, you have the next important ingredient to understand the structure of the ledger and transactions.
 
-A *transaction* is a list of *actions*, and there are just three kinds of action: ``create``, ``exercise`` and ``fetch``. All actions are performed on a contract.
+A *transaction* is a list of *actions*, and there are just four kinds of action: ``create``, ``exercise``, ``fetch`` and ``key assertion``.
 
-- A ``create`` action contains the contract arguments and changes the contract's status from *non-existent* to *active*.
-- A ``fetch`` action  checks the existence and activeness of a contract.
-- An ``exercise`` action contains the choice arguments and a transaction called the *consequences*. Exercises come in two kinds called ``consuming`` and ``nonconsuming``. ``consuming`` is the default kind and changes the contract's status from *active* to *archived*.
+- A ``create`` action creates a new contract with the given arguments and sets its status to *active*.
+- A ``fetch`` action checks the existence and activeness of a contract.
+- An ``exercise`` action exercises a choice on a contract resulting in a transaction (list of sub-actions) called the *consequences*. Exercises come in two kinds called ``consuming`` and ``nonconsuming``. ``consuming`` is the default kind and changes the contract's status from *active* to *archived*.
+- A ``key assertion`` records the assertion that the given contract key (see :ref:`contract_keys`) is not assigned to any active contract on the ledger.
 
-The consequences of exercise nodes turn each transaction into an ordered tree of (sub-) transactions, or, equivalently, a forest of actions. Actions are in one-to-one correspondence with proper sub-transactions. You can see the action and their consequences in the transaction view of the above scenario:
+Each action can be visualized as a tree, where the action is the root node, and its children are its consequences. Every consequence may have further consequences. As ``fetch``, ``create`` and ``key assertion`` actions have no consequences, they are leaf nodes. You can see the actions and their consequences in the transaction view of the above scenario:
```

Comment:
```suggestion
Each action can be visualized as a tree, where the action is the root node, and its children are its direct consequences. Every consequence may have further consequences. As ``fetch``, ``create`` and ``key assertion`` actions have no consequences, they are always leaf nodes. You can see the actions and their consequences in the transaction view of the above scenario:
```

Reply:
I'm including the `always` word.

However, I suggest to avoid the `direct` word, because the _consequenses_ are introduced as the (direct) children of the execution. So the implied _indirect consequences_ would require to change this definition (we can do this, but I intended this PR as a tidy-up only) here and on https://docs.daml.com/concepts/ledger-model/ledger-structure.html#actions-and-transactions  , where the transitive closure is called the set of _non-proper subactions_ instead.

Reply:
fine